build.rs

```diff
@@ -170,147 +204,149 @@ fn probe_r_paths() -> io::Result<InstallationPaths> {
-    }
-
-    if is_str_digit(&input[1]) {
-        result.minor = (&input[1]).to_string();
-    } else {
-        return Err(EnvVarError::InvalidEnvVar("Cannot find R minor version"))
-    }
-
-    if is_str_digit(&input[2]) {
-        result.patch = (&input[2]).to_string();
-    } else {
-        return Err(EnvVarError::InvalidEnvVar("Cannot find R patch level"))
-    }
+    // Split "<major>.<minor>.<patch>" to "<major>", "<minor>", and "<patch>"
+    let r_version_split = r_version
+        .split('.')
+        .map(|s| {
+            // Good:
+            //   - "4.1.2"
+            //
+            // Bad:
+            //   - "4.1.foo" (some part contains any non-digit characters)
+            //   - "4.1." (some part is missing)
+            if !s.is_empty() && s.chars().all(|c| c.is_digit(10)) {
+                Some(s)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<Option<&str>>>();
+
+    let (major, minor, patch) = match *r_version_split.as_slice() {
+        // if any of the first three items doesn't exist, the format is invalid
+        [] | [None, ..] => return Err(EnvVarError::InvalidEnvVar("Cannot find R major version")),
+        [_, None, ..] => return Err(EnvVarError::InvalidEnvVar("Cannot find R minor version")),
+        [_, _, None, ..] => return Err(EnvVarError::InvalidEnvVar("Cannot find R patch level")),
+        // if all of the first three items exist, the format is valid
+        [Some(major), Some(minor), Some(patch)] => {
+            (major.to_string(), minor.to_string(), patch.to_string())
+        }
+        // if the length is longer than 3, the format is invalid
+        _ => return Err(EnvVarError::InvalidEnvVar("Invalid format")),
+    };
 
-    if input.len() == 4 {
-        if input[3] == "devel" {
-            result.devel = "-devel".into();
-        } else {
-            return Err(EnvVarError::InvalidEnvVar("Cannot find R development status"));
+    let devel = match devel {
+        Some("devel") => true,
+        Some(_) => {
+            return Err(EnvVarError::InvalidEnvVar(
+                "Cannot find R development status",
+            ))
         }
-    }
+        None => false,
+    };
 
-    Ok(result)
+    Ok(RVersionInfo {
+        major,
+        minor,
+        patch,
+        devel,
+        version_string,
+    })
 }
 
-
-fn get_r_version_from_env(r_version_env_var:&str) -> Result<RVersionInfo, EnvVarError> {
+fn get_r_version_from_env(r_version_env_var: &str) -> Result<RVersionInfo, EnvVarError> {
     std::env::var(r_version_env_var)
         // Any error arising from reading env var is converted to this value
         .map_err(|_| EnvVarError::EnvVarNotPresent)
-        .map(|v| {
-            v.split(&['.', '-'][..])
-                .map(|s| s.to_string())
-                .collect::<Vec<String>>()
-        })
-        .and_then(str_vec_to_version)
+        .and_then(|s| parse_r_version(s, None))
```

Comment:
Do we need here `None`? Is this second parameter used anywhere else?

Reply:
Yeah, I too was wondering if I can remove `version_string` at all... Let's do it.